internal/instantRunoff/instantRunoff.go

```diff
@@ -13,15 +13,72 @@ type Vote struct {
 	RankedChoices []int
 }
 
-func Run(votes []*Vote, nameMap map[int]string) string {
+type Tally struct {
+	ID         int
+	Name       string
+	Round      int
+	Count      int
+	Eliminated bool
+	Winner     bool
+}
+
+type InstantRunoff struct {
+	Rounds  int
+	Tallies []*Tally
+	Winner  string
+}
+
+func (ir *InstantRunoff) ResultsAsString() string {
+	talliesByRound := make(map[int][]*Tally)
+
+	for _, tally := range ir.Tallies {
+		if v, ok := talliesByRound[tally.Round]; ok {
+			talliesByRound[tally.Round] = append(v, tally)
+			continue
+		}
+		talliesByRound[tally.Round] = []*Tally{tally}
+	}
+
+	var str strings.Builder
+
+	keys := make([]int, 0, len(talliesByRound))
+	for k := range talliesByRound {
+		keys = append(keys, k)
+	}
+	sort.Ints(keys)
+
+	for _, round := range keys {
+		tallies := talliesByRound[round]
+
+		str.WriteString("Round ")
+		str.WriteString(strconv.Itoa(round))
+		str.WriteString("\n\nCurrent first-choice votes:\n")
+
+		var eliminated []string
+		for _, tally := range tallies {
+			str.WriteString(fmt.Sprintf(" - %s: %d votes\n", tally.Name, tally.Count))
+
+			if tally.Eliminated {
+				eliminated = append(eliminated, tally.Name)
+			}
+		}
+		if len(eliminated) > 0 {
+			str.WriteString(fmt.Sprintf("\nEliminated %s\n\n---\n\n", strings.Join(eliminated, ", ")))
+		}
+	}
+
+	str.WriteString(fmt.Sprintf("\nThe winner of this election is %s", ir.Winner))
+
+	return str.String()
+}
+
+func Run(votes []*Vote, nameMap map[int]string) (*InstantRunoff, error) {
 	var ids []int
 	for k := range nameMap {
 		ids = append(ids, k)
 	}
 
-	var log strings.Builder
-
-	logFrequencies := func(freqs map[int]int) {
+	logFrequencies := func(freqs map[int]int) []*Tally {
```

Comment:
This isn't logging any frequencies now, it's just tallying votes so it's probably worth changing this function's name. Since it's also not messing with any variables in `Run`'s scope any more, it's worth pulling this whole function out into it's own top level thing.

Reply:
This is now `getTallies` and is a top level function